lib/Sema/TypeCheckAttr.cpp

```diff
@@ -3592,7 +3609,20 @@ DifferentiableAttributeParameterIndicesRequest::evaluate(
 }
 
 // SWIFT_ENABLE_TENSORFLOW
-void AttributeChecker::visitDerivativeAttr(DerivativeAttr *attr) {
+/// Typechecks the given derivative attribute `attr` on decl `D`.
+///
+/// Effects are:
+/// - Sets the parameter indices on `attr`.
+/// - Sets `attr` to invalid when there are errors.
+/// - If `AC` is non-null, diagnoses errors using the `AC` diagnostic methods.
+/// - Stores the attribute in the `ASTContext` list of derivative attributes.
+/// - Stores the derivative configuration in the original function's list of
+///   derivative configurations.
+static void typeCheckDerivativeAttr(ASTContext &Ctx, AttributeChecker *AC,
```

Comment:
I believe `AttributeChecker` is just a thin wrapper around an `ASTContext` and we can call `ASTContext::diagnose` instead. How about removing the `AttributeChecker *` argument?

If it's possible to use `DiagnosticTransaction` in callers to abort diagnostics instead of conditionally emitting diagnostics based on a flag, that would be preferable too!

Reply:
good idea, done